### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by uladzislau-zhurauski**

### CODE REVIEW

File: shop/controllers/image.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    @classmethod
+    def get_image(cls, image_pk):
+        try:
+            return ImageDAL.get_image_by_pk(image_pk)
+        except Image.DoesNotExist:
+            raise Http404
+
+    @classmethod
+    def update_image(cls, image_pk, image, content_type, object_id):
```
Comment (veranika-zakharchenia) [inline shop/controllers/image.py:34]: why do you get_image, after create_image, after you validate created image and all these actions take place in update method?

File: shop/controllers/product.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            else:
+                new_material = ProductMaterialDAL.insert_material(material_name)
+                ProductMaterialDAL.add_products(new_material, [product_obj])
+
+    @classmethod
+    def update_product(cls, product_pk, category, name, price, description, size, weight, stock, is_available,
+                       materials=None, images=None, images_to_delete=None):
+        product_obj = cls.get_product(product_pk, True)
+        cls.update_product_materials(product_obj, materials)
+        cls.process_images_to_delete(product_obj, images_to_delete)
```
Comment (veranika-zakharchenia) [inline shop/controllers/product.py:54]: why do you call images_to_delete when it is None?

File: shop/controllers/product_material.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    def update_material(cls, material_pk, name):
+        ProductMaterialDAL.update_material(cls.get_material(material_pk), name)
+
+    @classmethod
+    def delete_material(cls, material_pk):
+        ProductMaterialDAL.delete_material(cls.get_material(material_pk))
+
+    @classmethod
+    def is_the_same_name(cls, material_pk, new_name):
+        if new_name == cls.get_material(material_pk).name:
```
Comment (veranika-zakharchenia) [inline shop/controllers/product_material.py:35]:
```suggestion
        return new_name == cls.get_material(material_pk).name:
```

File: shop/permissions.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        owner_field_name = ''
+
+        def has_object_permission(self, request, view, obj):
+            return getattr(obj, self.owner_field_name) == request.user or request.user.is_staff
+
+    IsOwnerOrAdmin.owner_field_name = owner_field_name
+    return IsOwnerOrAdmin
+
+
+class IsUserOrAdmin(BasePermission):
```
Comment (veranika-zakharchenia) [inline shop/permissions.py:17]:
```suggestion
class PermissionValidator(BasePermission):
```
or checker

File: shop/controllers/product_material.py (hunk 3 above)
Comment (uladzislau-zhurauski) [reply shop/controllers/product_material.py:35]: Oops, really

File: shop/urls.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+urlpatterns = [
+    path('addresses/', AddressView.as_view(http_method_names=['get', 'post']), name='address-list'),
+    path('addresses/<int:pk>/', AddressView.as_view(http_method_names=['get', 'put', 'delete']), name='address-detail'),
+    path('auth/', include('rest_framework.urls', namespace='rest_framework')),
+    path('categories/', CategoryView.as_view(http_method_names=['get', 'post']), name='category-list'),
+    path('categories/<int:pk>/', CategoryView.as_view(http_method_names=['get', 'put', 'delete']),
+         name='category-detail'),
+    path('category/<int:category_pk>/', ProductView.as_view(http_method_names=['get']),
+         name='product-list-by-category'),
```
Comment (veranika-zakharchenia) [inline shop/urls.py:21]: what is the difference between these two endpoints?

File: shop/tests/test_views/test_order.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            (ClientType.AUTH_CLIENT, existent_pk, status.HTTP_403_FORBIDDEN),
+            (ClientType.AUTHOR_CLIENT, nonexistent_pk, status.HTTP_404_NOT_FOUND),
+            (ClientType.AUTHOR_CLIENT, existent_pk, status.HTTP_204_NO_CONTENT),
+            (ClientType.ADMIN_CLIENT, nonexistent_pk, status.HTTP_404_NOT_FOUND),
+            (ClientType.ADMIN_CLIENT, existent_pk, status.HTTP_204_NO_CONTENT)
+        ]
+    )
+    def test_delete_order(self, client_type, status_code, order_pk, multi_client):
+        url = reverse('order-detail', kwargs={'pk': order_pk})
+        if client_type == ClientType.AUTHOR_CLIENT and order_pk == existent_pk:
```
Comment (veranika-zakharchenia) [inline shop/tests/test_views/test_order.py:101]: separate test for each case

File: shop/tests/test_views/test_order.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        (ClientType.NOT_AUTH_CLIENT, status.HTTP_403_FORBIDDEN),
+        (ClientType.AUTH_CLIENT, status.HTTP_403_FORBIDDEN),
+        (ClientType.AUTHOR_CLIENT, status.HTTP_200_OK),
+        (ClientType.ADMIN_CLIENT, status.HTTP_200_OK)
+    ])
+    def test_put_order(self, client_type, status_code, multi_client):
+        order = Order.objects.first()
+        data = {'address': existent_pk}
+        url = reverse('order-detail', kwargs={'pk': order.pk})
+        user = order.user if client_type == ClientType.AUTHOR_CLIENT else None
```
Comment (veranika-zakharchenia) [inline shop/tests/test_views/test_order.py:84]: no conditions in tests

File: shop/tests/test_views/test_order.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+@pytest.mark.django_db
+class TestOrderViews:
+    @pytest.mark.parametrize('client_type, status_code', [
+        (ClientType.NOT_AUTH_CLIENT, status.HTTP_403_FORBIDDEN),
+        (ClientType.AUTH_CLIENT, status.HTTP_200_OK),
+        (ClientType.ADMIN_CLIENT, status.HTTP_200_OK)
+    ])
+    def test_get_order_list(self, client_type, status_code, multi_client):
+        if client_type == ClientType.AUTH_CLIENT:
```
Comment (veranika-zakharchenia) [inline shop/tests/test_views/test_order.py:20]: one test must describe one conception

File: shop/tests/test_views/test_image.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        data = {
+            'content_type': content_type,
+            'object_id': '' if object_id is None else object_id
+        }
+        if image == Arg.CORRECT:
+            data['image'] = get_in_memory_image_file
+        elif image == Arg.INCORRECT:
+            data['image'] = get_in_memory_file
+
+        if method_type == 'post':
```
Comment (veranika-zakharchenia) [inline shop/tests/test_views/test_image.py:62]: the same

File: shop/tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+register(FeedbackFactory)
+register(UserFactory)
+register(AddressFactory)
+register(ProductImageFactory)
+register(FeedbackImageFactory)
+register(OrderFactory)
+register(OrderItemFactory)
+register(ProductMaterialFactory)
+
+nonexistent_pk = 0
```
Comment (veranika-zakharchenia) [inline shop/tests/conftest.py:25]: consts should be in upper case

File: shop/serializers/image.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                    break
+            representation.pop(field_name)
+        return representation
+
+    @staticmethod
+    def get_content_object(instance):
+        for image_model_class in get_image_models():
+            if isinstance(instance.content_object, image_model_class):
+                image_model_module = importlib.import_module(f'{__package__}.{image_model_class.__name__.lower()}')
+                image_model_serializer = getattr(image_model_module, f'{image_model_class.__name__}OutputSerializer')
```
Comment (veranika-zakharchenia) [inline shop/serializers/image.py:34]: move class name  to variable

File: shop/urls.py (hunk 5 above)
Comment (uladzislau-zhurauski) [reply shop/urls.py:21]: `category/<int:category_pk>/` retrieves a list of products filtered by a specific category

**Review by veronica-zdanovich — APPROVED** [review]

**Pull request merged**